### PR TITLE
Remove concat, ++ and + overloads from MapView

### PR DESCRIPTION
**Pull request by joshlemer**
According to the [blog post about the Collections Rework](https://www.scala-lang.org/blog/2017/02/28/collections-rework.html), and I've heard it said quite a few times:

>Conceptually a `View` is now a reified operation over an `Iterator`.  

So from that, we should be able to codify this as roughly:

>A `View` transformation should be equivalent to the corresponding transformation on `Iterator`

e.g.
```scala
val iterable: Iterable[A] = ???
a.view.transform().toList == a.iterator.transform().toList
```
  
However, these `concat`/`++`/`+` method overloads on `MapView` violate this principle, by treating MapView more like a Map than like a View, in order to return a MapView rather than widening to generic View. The result is this:

```scala
scala> Map(1 -> 1).view.concat(Map(1 -> 2)).toList
res2: List[(Int, Int)] = List((1,2))

scala> Map(1 -> 1).iterator.concat(Map(1 -> 2)).toList
res3: List[(Int, Int)] = List((1,1), (1,2))
```

And, since it's a specialized overload, it returns different data depending on the static type of the argument or receiver:

```scala
scala> Map(1 -> 1).view.concat(Map(1 -> 2)).toList
res5: List[(Int, Int)] = List((1,2))

scala> Map(1 -> 1).view.concat(Map(1 -> 2) : Iterable[(Int, Int)]).toList
res6: List[(Int, Int)] = List((1,1), (1,2))

scala> (Map(1 -> 1) : Iterable[(Int, Int)]).view.concat(Map(1 -> 2)).toList
res7: List[(Int, Int)] = List((1,1), (1,2))
```

Which is pretty odd I reckon.

### CODE REVIEW

**Comment by julienrf** [comment]
To me, this is OK if a `MapView`’s iterator skips duplicate keys when it is traversed. That does not violate the “reified operations on `Iterator` definition” I think. But maybe a better way to think of this problem is to ask “is the current implementation a problem?”. If it is not, maybe the problem comes from the way we described `View` operations?

> And, since it's a specialized overload, it returns different data depending on the static type of the argument or receiver:
> 
> ```scala
> scala> Map(1 -> 1).view.concat(Map(1 -> 2)).toList
> res5: List[(Int, Int)] = List((1,2))
> 
> scala> Map(1 -> 1).view.concat(Map(1 -> 2) : Iterable[(Int, Int)]).toList
> res6: List[(Int, Int)] = List((1,1), (1,2))
> 
> scala> (Map(1 -> 1) : Iterable[(Int, Int)]).view.concat(Map(1 -> 2)).toList
> res7: List[(Int, Int)] = List((1,1), (1,2))
> ```
> 
> Which is pretty odd I reckon.

We should fix this problem by changing the signature of that overload as follows:

~~~ diff
- def concat[V1 >: V](that: SomeMapOps[K, V1]): MapView[K, V1] = new MapView.Concat(this, that)
+ def concat[V1 >: V](that: IterableOnce[(K, V1)]): MapView[K, V1] = new MapView.Concat(this, that)
~~~

**Comment by joshlemer** [comment]
> We should fix this problem by changing the signature of that overload as follows:
> 
> ```diff
> - def concat[V1 >: V](that: SomeMapOps[K, V1]): MapView[K, V1] = new MapView.Concat(this, that)
> + def concat[V1 >: V](that: IterableOnce[(K, V1)]): MapView[K, V1] = new MapView.Concat(this, that)
> ```

The issue with this approach is that in the general case a performant MapView cannot be constructed from a `left: MapView` and `right: IterableOnce[(K, V)]`. Just to perform a `get(key)` operation, you would need to force/traverse the entirety of `right`. And to perform `.take(3).toList` will still involve forcing the entire `right: IterableOnce` to make sure the first three keys are not overwritten. 

But even if they could be done performantly, I do think that the current overloads pose a problem because they are an inconsistency which harms simplicity. The major result of redesigning `Views` from the ground up is that `Views` are an independent collection type in their own right, with clear / consistent semantics independent of their source collection type (like in the old `View`s), similarly to java streams. But these overloads incompatibly modify `MapView` behaviour to behave like `Map`s rather than like `View`s, but in fact they are not `Map`s (according to the hierarchy, they extend View, not Map).

**Comment by julienrf** [comment]
> The issue with this approach is that in the general case a performant MapView cannot be constructed from a `left: MapView` and `right: IterableOnce[(K, V)]`. Just to perform a `get(key)` operation, you would need to force/traverse the entirety of `right`. And to perform `.take(3).toList` will still involve forcing the entire `right: IterableOnce` to make sure the first three keys are not overwritten.

That’s a good point.

> But even if they could be done performantly, I do think that the current overloads pose a problem because they are an inconsistency which harms simplicity.

Actually, yes. See also this confusing behavior:

~~~
(Set(1).view ++ Set(1)).foreach(println) // Prints “1” and “1”
(Map(1 -> 2).view ++ Map(1 -> 3)).foreach(println) // Only prints “1 -> 3”
~~~

File: src/library/scala/collection/MapView.scala
Comment (julienrf) [inline src/library/scala/collection/MapView.scala]: Is this change necessary?

**Review by julienrf — APPROVED** [review]

**Comment by julienrf** [comment]
You will have to whitelist your changes to MiMa.

File: src/library/scala/collection/MapView.scala
Comment (joshlemer) [reply src/library/scala/collection/MapView.scala]: Nope!

Comment (joshlemer) [reply src/library/scala/collection/MapView.scala]: (will remove)

**Comment by joshlemer** [comment]
@julienrf where can I learn about how to whitelist changes in MiMa?

**Comment by julienrf** [comment]
Sorry, I should have given you some pointers… The list of binary incompatibilities to exclude from MiMa is defined here:

https://github.com/scala/scala/blob/f921d0a0e7ccf498342b37f9b1a2d6dc2101fa47/build.sbt#L90-L93

It is almost empty right now because 2.13.0-RC2 has just been released.

I’m not sure why we don’t see a more detailed message in the CI, what do you see if you run `library/mimaReportBinaryIssues` in sbt? Then you’ll have to add your binary incompatibilities as explained in [MiMa’s documentation](https://github.com/lightbend/migration-manager#advanced-usage-filtering-binary-incompatibilities).

**Review by NthPortal — APPROVED** [review]

**Comment by adriaanm** [comment]
Generally, while we're in the RC cycle, please don't merge any PRs until I sign off on them. We're deciding today whether to do an RC3 and we really want the final to be out soon, so we have to be careful in what we include to avoid any regressions that could further delay the final.

**Comment by julienrf** [comment]
OK, sorry!

**Pull request merged**
